File: graphs/ltcd_841.py
'''
LeetCode - problem 841
'''

    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited_rooms = set()
        queue = collections.deque()
        queue.append(0)
        while queue:
            for _ in range(0, len(queue)):
                room_number = queue.popleft()
                if room_number not in visited_rooms:
                    found_keys = rooms[room_number]
                    for key in found_keys:
                        queue.append(key)
                    visited_rooms.add(room_number)
        if len(visited_rooms) == len(rooms):
            return True
        
        return False

File: graphs/ltcd_841.cpp
//LeetCode - problem 841

    vector<bool>visited;
    
    void dfs(vector<vector<int>>& rooms, int v) {
        
        if (visited[v]) return;
        
        visited[v] = true;
        for (auto it = rooms[v].begin(); it != rooms[v].end(); it++)
            dfs(rooms, *it);
    }
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        
        int n = rooms.size();
        
        visited.resize(n, false);
        
        dfs(rooms, 0);
        
        for (int i = 0; i < visited.size(); i++) 
            if (!visited[i])
                return false;
        
        return true;
    }
